File: AlgorithmTraining/AlgorithmTraining/d4/P1486.cpp
//
//  P1486.cpp
//  AlgorithmTraining
//
//  Created by 유태우 on 18/09/2018.
//  Copyright © 2018 유태우. All rights reserved.
//

#include <iostream>

#define MIN(a, b) (a < b) ? (a) : (b)

using namespace std;

int N, B;
int *H;

bool * visited;
int ans;

void DFS(int sum, int index){
    
    if(sum > ans){
        return;
    }
    
    if(sum >= B){
        ans = MIN(sum, ans);
        return;
    }
    
    for(int i=index; i<N; i++){
        if(!visited[i]){
            visited[i] = true;
            sum += H[i];
            DFS(sum, i+1);
            sum -= H[i];
            visited[i] = false;
        }
    }
    
}

int main(int argc, const char * argv[]) {
    
    int T;
    cin >> T;
    
    for(int tc=1; tc<=T; tc++){
        cin >> N >> B;
        H = new int[N];
        visited = new bool[N];
        ans = 2000000000;
        
        for(int i=0; i<N; i++){
            cin >> H[i];
        }
        
        DFS(0, 0);
        
        cout << "#" << tc << " " << ans-B << "\n";
        
    }
    
    return 0;
}

File: src/training2/d4/P1486.java
package training2.d4;

import java.util.Scanner;

public class P1486 {

    public static int N, B;
    public static int [] H;
    public static boolean [] visited;
    public static int ans;

    public static void main(String [] args){

        Scanner scn = new Scanner(System.in);

        int T = scn.nextInt();

        for(int tc=1; tc<=T; tc++){

            ans = Integer.MAX_VALUE-1;

            N = scn.nextInt();
            B = scn.nextInt();

            H = new int[N];
            visited = new boolean[N];
            for(int i=0; i<N; i++){
                H[i] = scn.nextInt();
            }

            DFS(0, 0);

            System.out.println(String.format("#%d %d", tc, ans-B));

        }

    }

    public static void DFS(int sum, int index){

        if(sum > ans){
            return;
        }

        if(sum >= B){
            ans = Math.min(sum, ans);
        }

        for(int i=index; i<N; i++){
            if(!visited[i]){
                sum += H[i];
                visited[i] = true;
                DFS(sum, i+1);
                visited[i] = false;
                sum -= H[i];
            }
        }

    }


}
